### PR TITLE
Swap GW2 API implementation and create API facade.

### PR DESCRIPTION
**Pull request by Xyaren**
Allows concurrent api requests, prevents API Key bleeding.
Added facade Tests.

### CODE REVIEW

**Comment by Xyaren** [comment]
Probably Fixes #35

**Comment by Xyaren** [comment]
It's worth mentioning, that the used api wrapper https://github.com/JuxhinDB/gw2-api-interface is the only wrapper listed as **active** on the GW2 Wiki Page: https://wiki.guildwars2.com/wiki/API:List_of_wrappers

File: bot/TS3Bot.py [hunk 1]
```diff
@@ -331,7 +331,7 @@ def setResetroster(self, date, red=[], green=[], blue=[], ebg=[]):
     def getActiveCommanders(self):
         return self.commander_checker.execute()
 
-    def removeGuild(self, name):
+    def removeGuild(self, name, tag=None):
```
Comment (ogrady) [inline bot/TS3Bot.py:334]: why is tag an optional parameter here? Should we not be able to determine the tag based on the name from the API in any case?

File: bot/rest/server.py [hunk 2]
```diff
@@ -36,14 +38,25 @@ def stop(self):
...
     ]
-
     for ctrl in controller:
         app.register_blueprint(ctrl.api)
-    return app
+
+
+def register_error_handlers(flask: Flask):
+    @flask.errorhandler(HTTPException)
+    def _handle_error(e: HTTPException):
```
Comment (ogrady) [inline bot/rest/server.py:61]: this smells **a lot** like creating control flow through the use of exceptions, which becomes visible in `_createGuild` of the `guild_controller`. The werkzeug module seems to facilitate this kind of control flow by offering exceptions for specific HTTP-codes and global exception handlers, but beside that, it is considered an anti-pattern.

File: bot/rest/controller/guild_controller.py [hunk 3]
```diff
@@ -26,15 +26,25 @@ def _create_guild():
...
+                raise BadRequest("Property $.name is required")
+            if tag is None:
+                raise BadRequest("Property $.tag is required")
+
+            res = self._bot.createGuild(name, tag, groupname, contacts)
+
+            if res == 0:
+                return jsonify("OK")
+            else:
+                raise BadRequest(f"Operation was not successful. Response code: {res}")
```
Comment (ogrady) [inline bot/rest/controller/guild_controller.py:39]: see https://github.com/FlussuferOrga/ts-gw2-verifyBot/pull/36/files#r469156113

File: bot/TS3Bot.py (hunk 1 above)
Comment (Xyaren) [reply bot/TS3Bot.py:334]: True, but Channels can/will be created with custom Tags.
The created channel thereby does not have the "official" Guild tag, and can not be found for deletion.

File: bot/rest/server.py (hunk 2 above)
Comment (Xyaren) [reply bot/rest/server.py:61]: This error handler just ensures, that any response will be a valid Json, so clients do not have to deal with different content types.

File: bot/TS3Bot.py (hunk 1 above)
Comment (ogrady) [reply bot/TS3Bot.py:334]: The name of the channels should always use the actual ingame-tag. The custom tags are just meant to deal with duplicate guild-groups in TS and should strictly be used as such. They can be looked up by the guild name in the `guilds`-table of the db:

```sql
SELECT ts_group FROM guilds WHERE guild_name = 'Unexpected Chaos'
```
Should yield `uC.`.

File: bot/rest/server.py (hunk 2 above)
Comment (ogrady) [reply bot/rest/server.py:61]: That is of course a fine addition, but in `bot/rest/controller/guild_controller.py` we have the exact situation where either a `jsonify`'d response is returned, or a `BadRequest`-error is raised. This makes the path the code takes more opaque, since it may bypass a caller in case of the thrown error while just bubbling all the way up to the set error-handler. This is not an issue in our case _yet_, but an anti-pattern nonetheless.

Comment (Xyaren) [reply bot/rest/server.py:61]: Hm, the used "abort(int,str)" function basically does the same, even sais it in the docs: 
`Raises an :py:exc: HTTPException for the given status code or WSGI
    application.`

File: bot/TS3Bot.py (hunk 1 above)
Comment (Xyaren) [reply bot/TS3Bot.py:334]: The tag provided on the create command is uses for the channel name: bot/TS3Bot.py:424
On delete the tag will be fetched from the api, this prevents deletion of the channel if a custom tag was used on create.

Comment (ogrady) [reply bot/TS3Bot.py:334]: That is working as intended. The custom tag you are referring to is the parameter `groupname` of the same function. That one is indeed optional, as it is the same as `tag` in any case where the tag of a guild can serve as a unique group name. The parameter `tag` should always hold the actual tag of the guild. It is only a parameter because no lookup on the API is done so far, but that is an outstanding issue https://github.com/FlussuferOrga/ts-gw2-verifyBot/issues/32
We could very well look up the tag from the API and drop the `tag`-parameter alltogether from `createGuild`.

File: bot/rest/server.py (hunk 2 above)
Comment (ogrady) [reply bot/rest/server.py:61]: Absolutely! It is recommended behaviour by the authors of the library; this is an issue in how the library itself encourages the handling, which I think is bad style in general. As I said, this is probably not a problem for our usecase yet, but I always see exceptions as `goto`s in disguise and I am therefore very sceptical when they are used to create control flow which basically could circumvent cleanup-code in callers or later segments of the function the exception is thrown in. Similar goes for `break`, `continue`, and `return` in the middle of a function, btw.

I would have stuck with returning a code and a message to the caller and have them create the appropriate HTTP response, but I am also not too eager on changing it back, if you are convinced by the current implementation.

Comment (Xyaren) [reply bot/rest/server.py:61]: I can see the issue since Python does not have checked Exceptions. On the other hand this makes return values of functions much more complicates, as basically every function would have to return some kind of error-situation response next to it's result.

Python itself seems to be using Exceptions even in "expected" flow: https://jeffknupp.com/blog/2013/02/06/write-cleaner-python-use-exceptions/

Exceptions are **meant to be handled** by the caller, as different callers might have different ways of dealing with a specific error, or even errors in general.
A webserver will return an error page, while a command line application will exit with some specific error code.
In the case of flask, you just raise a BadRequest Exception and let the error handler deal with it.
This also prevents lots of dublicated code, as you don't need to create a bunch of ErrorResponses in each controller.

To summarize: I agree on Exceptions beeing antipattern as beeing used for the expected control flow (e.g. "NoResultsFoundException" -> use None or an emtpy list as return) **but** for "Dead Ends" in the application, you do not want to deal with passing the error upwards with each method call in the stack on your own. That's what exceptions are meant to do.

File: bot/TS3Bot.py (hunk 1 above)
Comment (Xyaren) [reply bot/TS3Bot.py:334]: groupname and tag can be different.
```
POST /guild
{
  "name": "Stardust Alliance",
  "tag": "Durst",
  "groupname": "DUST,
  "contacts": [
    "UserName.1234"
  ]
}
```
Will create the channel `Stardust Alliance [Durst]`

```
DELETE /guild
{
  "name": "Stardust Alliance"
}
```
Does **not** delete the channel.

File: bot/rest/server.py (hunk 2 above)
Comment (ogrady) [reply bot/rest/server.py:61]: I agree on the "dead ends" situation, but I don't believe this is a dead end as failing when trying to create a guild is kind of an expected outcome, not a critical error. But I think in the end this may be an issue with my personal reservations against unstructured programming, which exceptions are inherently a part of, so I may be biased and the approach you suggested is probably fine as is.

File: bot/TS3Bot.py (hunk 1 above)
Comment (Xyaren) [reply bot/TS3Bot.py:334]: So dropping the Tag parameter on create and delete is the way to go?

Comment (ogrady) [reply bot/TS3Bot.py:334]: That is correct, there are no checks in place to ensure the tag that is being passed is the actual tag used in game. The intended use is for the caller to make sure name and tag are correct and to provide an optional groupname in case a group of that name already exists. This is by no means optimal, especially since deleting a guild causes issues if someone accidentally created a guildchannel for a guild that does not exist, which is checked upon deletion, but not during creation. That simply happend due to the feature being rushed. Dropping the tag parameter in both functions and looking it up from the API in both cases as you suggested is indeed the way to go. That would be a possible solution for #32 .

**Comment by Xyaren** [comment]
Removed the "tag" parameter from the guild endpoints.
Refactored guild channel functionality to GuildService.
